core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala

```diff
@@ -110,6 +110,12 @@ private[spark] class UnifiedMemoryManager private[memory] (
       evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = synchronized {
     assert(onHeapExecutionMemoryPool.poolSize + storageMemoryPool.poolSize == maxMemory)
     assert(numBytes >= 0)
+    if (numBytes > maxMemory) {
+      // Fail fast if the block simply won't fit
+      logInfo(s"Will not store $blockId as the required space ($numBytes bytes) exceeds our " +
+        s"memory limit ($maxMemory bytes)")
```

Comment:
this check is actually duplicated now that you've removed it from `ensureFreeSpace`, but I can see that it stops us from even resizing the memory pools early so it's OK to keep here